skyrl-train/tests/cpu/test_trainer_utils.py

```diff
@@ -321,6 +325,313 @@ def test_dump_per_dataset_eval_results_comprehensive(mock_file):
             continue
 
 
+def test_handle_dynamic_sampling_null_strategy():
+    """Test that null strategy returns input unchanged."""
+    generator_output = {
+        "prompt_token_ids": [[1, 2, 3], [4, 5, 6]],
+        "response_ids": [[7, 8], [9, 10]],
+        "rewards": [[1.0, 2.0], [3.0, 4.0]],
+        "loss_masks": [[1, 1], [1, 1]],
+        "stop_reasons": ["stop", "stop"],
+        "rollout_metrics": None,
+    }
+    uids = ["uid1", "uid2"]
+    sampling_config = {"type": None}
+
+    result_output, result_uids, keep_sampling, state = handle_dynamic_sampling(generator_output, uids, sampling_config)
+
+    assert result_output == generator_output
+    assert result_uids == uids
+    assert keep_sampling is False
+    assert state is None
+
+
+def test_handle_dynamic_sampling_invalid_strategy():
+    """Test that invalid strategy raises ValueError."""
+    generator_output = {
+        "prompt_token_ids": [[1, 2, 3]],
+        "response_ids": [[7, 8]],
+        "rewards": [[1.0, 2.0]],
+        "loss_masks": [[1, 1]],
+    }
+    uids = ["uid1"]
+    sampling_config = {"type": "invalid_strategy"}
+
+    with pytest.raises(ValueError, match="Invalid dynamic sampling type: invalid_strategy"):
+        handle_dynamic_sampling(generator_output, uids, sampling_config)
+
+
+def test_handle_replace_sampling_sufficient_good_samples():
+    """Test replace sampling when there are sufficient good samples (>0.3)."""
+    # Create test data with some good UIDs (high variance) and some bad UIDs (zero variance)
+    generator_output = {
+        "prompt_token_ids": [[1, 2], [3, 4], [5, 6], [7, 8], [9, 10], [11, 12]],
```

Comment:
```suggestion
        "prompt_token_ids": [[1, 2], [1, 2], [3, 4], [3, 4], [5, 6], [5, 6]],
```

Nit: this list should match the repetition pattern in `uids`